src/mono/msbuild/WasmAppBuilder/WasmAppBuilder.cs

```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class WasmAppBuilder : Task
+{
+    public ITaskItem[] PInvokeModules { get; set; }
+    public ITaskItem[] PInvokeAssemblies { get; set; }
+
+	public string PInvokeTablePath { get; set; }
+
+	public override bool Execute () {
+		if (PInvokeTablePath == null)
+            throw new ArgumentException("Only PInvokeTablePath=true is supported right now.");
+
+		GenPInvokeTable (PInvokeModules.Select (item => item.ItemSpec).ToArray (), PInvokeAssemblies.Select (item => item.ItemSpec).ToArray ());
+
+		return true;
+	}
+
+	void GenPInvokeTable (string[] pinvokeModules, string[] assemblies) {
+		var modules = new Dictionary<string, string> ();
+		foreach (var module in pinvokeModules)
+			modules [module] = module;
+
+		var pinvokes = new List<PInvoke> ();
+
+		var resolver = new PathAssemblyResolver (assemblies);
+		var mlc = new MetadataLoadContext (resolver, "System.Private.CoreLib");
```

Comment:
Why only SPC? Any src/library can include PInvoke?

Reply:
Will add more libraries as needed. Running this is required to create the final executable, running it during the build would mean the executable needs to be recompiled/relinked, which takes a long time and requires emscripten.

Reply:
wasm doesn't support dlopen, so these mapping tables need to be compiled into the executable. There are two modes right now:
- during development, we use a precompiled dotnet.wasm created here.
- for release, the same can be used, or its possible to recompile/relink it to make it smaller.

Reply:
Right, so we should run this for all 100 or so assemblies which are part of netcore as any of them can include pinvoke call.